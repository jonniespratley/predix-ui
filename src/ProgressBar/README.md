The `<ProgressBar/>` component enables you to use the design system progress bar.


## Usage

```react
const { ProgressBar } = PxReact;
<div>
  <ProgressBar value={75}/>
</div>
```

### Infinite example

```react
const { ProgressBar } = PxReact;
<div>
  <ProgressBar value={75} infinite/>
</div>
```


## Properties

```table
span: 6
rows:
  - Name: infinite
    Type: bool
    Description: Flag for whether to animate the progress bar. Set to true for an indeterminate or infinite progress bar.
  - Name: value
    Type: number
    Description: Represents the value (from 0 to 100) of the percentage progress.
```


### Styling

```
Custom property | Description | Default

`--px-progress-bar-height` | Height of the progress bar element | `5px`
`--px-progress-bar-fill-color` | Color of the filled portion of the bar | black
`--px-progress-bar-background-color` | Color of the unfilled portion of the bar | lightgray
```
