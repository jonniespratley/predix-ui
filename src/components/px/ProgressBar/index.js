import React from 'react';
import PropTypes from 'prop-types';
import styled, { css, keyframes } from 'styled-components';

const ProgressAnimation = keyframes`
  0% {
    left     : 0;
    right    : 100%;
    transform: scaleX(0);
  }
  50% {
    left     : 0;
    right    : 0;
    transform: scaleX(1);
  }
  100% {
    left     : 100%;
    right    : 0;
    transform: scaleX(0);
  }
`;

/* eslint-disable-next-line */
const _computeRatio = v => (v < 0 ? 0 : (v > 100 ? 1 : v / 100));

const ProgressFill = styled.div`
  background-color        : var(--px-progress-bar-fill-color,black);
  height                  : var(--px-progress-bar-height, 5px);
  transition              : transform 1s cubic-bezier(.78,.13,.16,.87);
  -webkit-transform-origin: left center;
  transform-origin        : left center;
  -webkit-transform       : scaleX(0);
  transform               : scaleX(0);
  overflow                : hidden;
  
  ${props => props.value && css`
    transform               : scaleX(${_computeRatio(props.value)});
  `}
  ${props => props.infinite && css`
    position : relative;
    animation: ${ProgressAnimation} 2s cubic-bezier(.78,.13,.16,.87) infinite;
  `}
`;
const ProgressBackground = styled.div`
  background-color: var(--px-progress-bar-background-color, lightgray);
  height          : var(--px-progress-bar-height, 5px);
`;

const ProgressBar = styled.div`
  height: var(--px-progress-bar-height, 5px);
`;

const Component = ({
  value,
  max,
  min,
  infinite
}) => (
  <div aria-valuemin={min} aria-valuemax={max}>
    <ProgressBar>
      <ProgressBackground>
        <ProgressFill value={value} infinite={infinite} />
      </ProgressBackground>
    </ProgressBar>
  </div>
);

Component.defaultProps = {
  value: 0,
  max: 100,
  min: 0,
  infinite: false
};
Component.propTypes = {
  value: PropTypes.number,
  max: PropTypes.number,
  min: PropTypes.number,
  infinite: PropTypes.bool
};

export default Component;
