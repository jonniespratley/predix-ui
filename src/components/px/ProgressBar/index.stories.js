import React from 'react';
import { storiesOf } from '@storybook/react';
import ProgressBar from './';
///
storiesOf('Progress Bar', module)
	.add('default', () => (
		<ProgressBar value={75}/>
	))
	.add('with infinite', () => (
		<ProgressBar value={75} infinite/>
	));
