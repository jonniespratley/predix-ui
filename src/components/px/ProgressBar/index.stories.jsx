import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, number } from '@storybook/addon-knobs/react';
import ProgressBar from './';

storiesOf('Components / Progress Bar', module)
  .addDecorator(withKnobs)
  .add('default', () => (
    <ProgressBar
      value={number('value', 50)}
      infinite={boolean('infinite', false)}
    />
  ));
