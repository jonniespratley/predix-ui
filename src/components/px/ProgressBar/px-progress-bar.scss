/**
 * px-progress-bar styles
 */
.px-progress-bar {
  // Component

  $px-progress-bar-fill-color      : var(--px-progress-bar-fill-color,black);
  $px-progress-bar-background-color: var(--px-progress-bar-background-color, lightgray);
  $px-progress-bar-height          : var(--px-progress-bar-height, 5px);
  @keyframes progress {
    0% {
      left     : 0;
      right    : 100%;
      transform: scaleX(0);
    }
    50% {
      left     : 0;
      right    : 0;
      transform: scaleX(1);
    }
    100% {
      left     : 100%;
      right    : 0;
      transform: scaleX(0);
    }
  }

  height: $px-progress-bar-height;

  .background {
    background-color: $px-progress-bar-background-color;
    height          : $px-progress-bar-height;
  }

  .fill {
    background-color        : $px-progress-bar-fill-color;
    height                  : $px-progress-bar-height;
    transition              : transform 1s cubic-bezier(.78,.13,.16,.87);
    -webkit-transform-origin: left center;
    transform-origin        : left center;
    -webkit-transform       : scaleX(0);
    transform               : scaleX(0);
    overflow                : hidden;
  }

  &.infinite {
    .fill {
      position : relative;
      animation: progress 2s cubic-bezier(.78,.13,.16,.87) infinite;
    }
  }


}
