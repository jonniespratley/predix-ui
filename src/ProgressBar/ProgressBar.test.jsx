import React from 'react';
import renderer from 'react-test-renderer';
import ProgressBar from '.';

describe('ProgressBar', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<ProgressBar value={50} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('infinite - renders correctly', () => {
    const tree = renderer.create(<ProgressBar infinite value={50} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
