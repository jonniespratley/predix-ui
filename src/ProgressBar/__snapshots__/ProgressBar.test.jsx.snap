// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar infinite - renders correctly 1`] = `
.c2 {
  background-color: var(--px-progress-bar-fill-color,black);
  height: var(--px-progress-bar-height,5px);
  -webkit-transition: -webkit-transform 1s cubic-bezier(0.78,0.13,0.16,0.87);
  -webkit-transition: transform 1s cubic-bezier(0.78,0.13,0.16,0.87);
  transition: transform 1s cubic-bezier(0.78,0.13,0.16,0.87);
  -webkit-transform-origin: left center;
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scaleX(0);
  -webkit-transform: scaleX(0);
  -ms-transform: scaleX(0);
  transform: scaleX(0);
  overflow: hidden;
  -webkit-transform: scaleX(0.5);
  -ms-transform: scaleX(0.5);
  transform: scaleX(0.5);
  position: relative;
  -webkit-animation: gxHwZO 2s cubic-bezier(0.78,0.13,0.16,0.87) infinite;
  animation: gxHwZO 2s cubic-bezier(0.78,0.13,0.16,0.87) infinite;
}

.c1 {
  background-color: var(--px-progress-bar-background-color,lightgray);
  height: var(--px-progress-bar-height,5px);
}

.c0 {
  height: var(--px-progress-bar-height,5px);
}

<div
  aria-valuemax={100}
  aria-valuemin={0}
>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      <div
        className="c2"
        value={50}
      />
    </div>
  </div>
</div>
`;

exports[`ProgressBar renders correctly 1`] = `
.c2 {
  background-color: var(--px-progress-bar-fill-color,black);
  height: var(--px-progress-bar-height,5px);
  -webkit-transition: -webkit-transform 1s cubic-bezier(0.78,0.13,0.16,0.87);
  -webkit-transition: transform 1s cubic-bezier(0.78,0.13,0.16,0.87);
  transition: transform 1s cubic-bezier(0.78,0.13,0.16,0.87);
  -webkit-transform-origin: left center;
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scaleX(0);
  -webkit-transform: scaleX(0);
  -ms-transform: scaleX(0);
  transform: scaleX(0);
  overflow: hidden;
  -webkit-transform: scaleX(0.5);
  -ms-transform: scaleX(0.5);
  transform: scaleX(0.5);
}

.c1 {
  background-color: var(--px-progress-bar-background-color,lightgray);
  height: var(--px-progress-bar-height,5px);
}

.c0 {
  height: var(--px-progress-bar-height,5px);
}

<div
  aria-valuemax={100}
  aria-valuemin={0}
>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      <div
        className="c2"
        value={50}
      />
    </div>
  </div>
</div>
`;
